File: src/main/java/com/dogGetDrunk/meetjyou/jwt/JwtManager.java
package com.dogGetDrunk.meetjyou.jwt;

import com.dogGetDrunk.meetjyou.common.exception.business.IncorrectJwtSubjectException;
import io.jsonwebtoken.Jwts;
import io.jsonwebtoken.security.Keys;
import org.springframework.beans.factory.annotation.Value;
import org.springframework.stereotype.Component;

import javax.crypto.SecretKey;
import java.util.Date;

@Component
public class JwtManager {

    private final SecretKey secretKey;
    @Value("${jwt.issuer}")
    private String issuer;
    @Value("${jwt.access-expiration}")
    private Long accessTokenExpiresIn;
    @Value("${jwt.refresh-expiration}")
    private Long refreshTokenExpiresIn;

    public JwtManager(@Value("${jwt.secret-key}") String secretKey) {
        this.secretKey = Keys.hmacShaKeyFor(secretKey.getBytes());
    }

    public String generateAccessToken(Long userId) {
        Date now = new Date();
        return Jwts.builder()
                .header()
                .type("JWT")
                .and()
                .issuer(issuer)
                .subject(Long.toString(userId))
                .issuedAt(now)
                .expiration(new Date(now.getTime() + accessTokenExpiresIn))
                .signWith(secretKey)
                .compact();
    }

    public String generateRefreshToken(Long userId) {
        Date now = new Date();
        return Jwts.builder()
                .header()
                .type("JWT")
                .and()
                .issuer(issuer)
                .subject(Long.toString(userId))
                .issuedAt(now)
                .expiration(new Date(now.getTime() + refreshTokenExpiresIn))
                .signWith(secretKey)
                .compact();
    }

    public void validateToken(String token, Long userId) {
        String userIdInToken = Jwts.parser()
                .verifyWith(secretKey)
                .build()
                .parseSignedClaims(token)
                .getPayload()
                .getSubject();

        if (!userIdInToken.equals(Long.toString(userId))) {
            throw new IncorrectJwtSubjectException(userId);
        }
    }
}

File: src/main/java/com/dogGetDrunk/meetjyou/jwt/JwtManager.kt
package com.dogGetDrunk.meetjyou.jwt

import com.dogGetDrunk.meetjyou.common.exception.business.jwt.IncorrectJwtSubjectException
import io.jsonwebtoken.Jwts
import io.jsonwebtoken.security.Keys
import org.springframework.beans.factory.annotation.Value
import org.springframework.stereotype.Component
import java.util.Date
import java.util.UUID
import javax.crypto.SecretKey

@Component
class JwtManager(
    @Value("\${jwt.secret-key}")
    secretKey: String,

    @Value("\${jwt.issuer}")
    private val issuer: String,

    @Value("\${jwt.access-expiration}")
    private val accessTokenExpiresIn: Long,

    @Value("\${jwt.refresh-expiration}")
    private val refreshTokenExpiresIn: Long
) {
    private val secretKey: SecretKey = Keys.hmacShaKeyFor(secretKey.toByteArray())

    fun generateAccessToken(uuid: UUID): String {
        val now = Date()
        return Jwts.builder()
            .header()
            .type("JWT")
            .and()
            .issuer(issuer)
            .subject(uuid.toString())
            .issuedAt(now)
            .expiration(Date(now.time + accessTokenExpiresIn))
            .signWith(secretKey)
            .compact()
    }

    fun generateRefreshToken(uuid: UUID): String {
        val now = Date()
        return Jwts.builder()
            .header()
            .type("JWT")
            .and()
            .issuer(issuer)
            .subject(uuid.toString())
            .issuedAt(now)
            .expiration(Date(now.time + refreshTokenExpiresIn))
            .signWith(secretKey)
            .compact()
    }

    fun validateToken(token: String?, uuid: UUID) {
        val userIdInToken = Jwts.parser()
            .verifyWith(secretKey)
            .build()
            .parseSignedClaims(token)
            .payload
            .subject

        if (userIdInToken != uuid.toString()) {
            throw IncorrectJwtSubjectException(uuid)
        }
    }
}
